File: examples/vsliceJava/src/test/java/csw/examples/vsliceJava/assembly/AssemblyTestData.java
package csw.examples.vsliceJava.assembly;

import akka.japi.Pair;
import csw.examples.vsliceJava.assembly.AssemblyContext.TromboneCalculationConfig;
import csw.examples.vsliceJava.assembly.AssemblyContext.TromboneControlConfig;
import csw.services.loc.ComponentId;
import csw.services.loc.Connection;
import csw.services.pkg.Component.AssemblyInfo;
import javacsw.services.loc.JComponentType;
import javacsw.services.pkg.JComponent;
import net.logstash.logback.encoder.org.apache.commons.lang.ArrayUtils;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.List;
import java.util.stream.Collectors;

import static javacsw.services.loc.JConnectionType.AkkaType;
import static csw.examples.vsliceJava.assembly.Algorithms.*;
import static javacsw.services.pkg.JComponent.RegisterAndTrackServices;
import static javacsw.util.JUtils.*;

@SuppressWarnings({"unused", "WeakerAccess"})
public class AssemblyTestData {

  static List<Double> testFocusErrors = Arrays.asList(ArrayUtils.toObject(
    new double[]{-20.0, -16.0, -12.0, -8.0, -4.0, 0.0, 4.0, 8.0, 12.0, 16.0, 20.0}));
  static List<Double> testZenithAngles = Arrays.asList(ArrayUtils.toObject(
    new double[]{0.0, 5.0, 10.0, 15.0, 20.0, 25.0, 30.0, 35.0, 40.0, 45.0, 50.0, 55.0, 60.0}));

  static ComponentId hcdId = new ComponentId("lgsTromboneHCD", JComponentType.HCD);

  static AssemblyInfo TestAssemblyInfo = JComponent.assemblyInfo(
    "lgsTrombone",
    "nfiraos.ncc.trombone",
    "csw.examples.vsliceJava.assembly.TromboneAssembly",
    RegisterAndTrackServices, Collections.singleton(AkkaType), Collections.singleton(new Connection.AkkaConnection(hcdId)));

  static TromboneCalculationConfig TestCalculationConfig = new TromboneCalculationConfig(
    95.0, .75, 20.0, -20.0, 4);

  static TromboneControlConfig TestControlConfig = new TromboneControlConfig(
    8.0, 225, 90.0, 200, 1200);

  static AssemblyContext TestAssemblyContext = new AssemblyContext(TestAssemblyInfo, TestCalculationConfig, TestControlConfig);

  static double maxRDError = focusErrorToRangeError(TestCalculationConfig, TestCalculationConfig.upperFocusLimit);
  static double maxRD = zenithAngleToRangeDistance(TestCalculationConfig.defaultInitialElevation, testZenithAngles.get(testZenithAngles.size() - 1));
  static double maxTotalRD = maxRDError + zenithAngleToRangeDistance(TestCalculationConfig.defaultInitialElevation, testZenithAngles.get(testZenithAngles.size() - 1));

  static double minRDError = focusErrorToRangeError(TestCalculationConfig, TestCalculationConfig.lowerFocusLimit);
  static double minRD = zenithAngleToRangeDistance(TestCalculationConfig.defaultInitialElevation, testZenithAngles.get(0));
  static double minTotalRD = minRDError + zenithAngleToRangeDistance(TestCalculationConfig.defaultInitialElevation, testZenithAngles.get(0));

  static double minReasonableAltitude = 85.0;
  static double maxReasonableAltitude = 110.0;
  static double minReasonableRange = minReasonableAltitude;
  static double maxReasonableRange = 200.0;
  static double minReasonableStage = minReasonableAltitude;
  static double maxReasonableStage = 200.0;
  static double maxReasonableRangeError = 5.0;
  static double minReasonableRangeError = -5.0;

  // These are values for testing the range error based on focus error (fe, fe distance error)
  static List<Pair<Double, Double>> feRangeErrorTestValues = testFocusErrors.stream().map(f ->
    new Pair<>(f, focusErrorToRangeError(TestCalculationConfig, f)))
    .collect(Collectors.toList());

  // This provides "base" range distance based on za only (za, rd)
  static List<Pair<Double, Double>> zaRangeDistance = testZenithAngles.stream().map(f ->
    new Pair<>(f, zenithAngleToRangeDistance(TestCalculationConfig.defaultInitialElevation, f)))
    .collect(Collectors.toList());

  // This produces an Vector of data of form (newRangeDistance, newElevation), (newRangeDistance, newElevation),...) as a function of za at a given focus error
  static List<Pair<Double, Double>> newRangeAndElData(Double fe) {
    return testZenithAngles.stream().map(f ->
      focusZenithAngleToElevationAndRangeDistance(TestCalculationConfig, TestCalculationConfig.defaultInitialElevation, fe, f))
      .collect(Collectors.toList());
  }

  // This produces an Vector of data of form (newRangeDistance, newElevation), (newRangeDistance, newElevation),...) as a function of za at a given focus error
  static List<Pair<Double, Double>> newZARangeAndElData(Double za) {
    return testFocusErrors.stream().map(f ->
      focusZenithAngleToElevationAndRangeDistance(TestCalculationConfig, TestCalculationConfig.defaultInitialElevation, f, za))
      .collect(Collectors.toList());
  }

  static class TestValue {
    final Pair<Pair<Double, Double>, Integer> v;

    public TestValue(Pair<Pair<Double, Double>, Integer> v) {
      this.v = v;
    }
  }

  static double getza(TestValue t) {
    return t.v.first().first();
  }

  static double gettrd(TestValue t) {
    return t.v.first().second();
  }

  static int getenc(TestValue t) {
    return t.v.second();
  }


  /**
   * This routine takes the two configs and returns a tuple of ((zenith angle, totalRangeDistance), encoder position)
   * for the range of zenith angle 0 to 60 in steps of 5 degrees
   *
   * @param calculationConfig the assembly calculation config provides algorithm parameters
   * @param controlConfig     the control config provides values for mapping stage positions into encoder values
   * @return a Vector of the form (((za, totalRD), enc), ((za, totalRD), enc), ...)
   */
  static List<TestValue> calculatedTestData(TromboneCalculationConfig calculationConfig, TromboneControlConfig controlConfig, double fe) {

    double rde = focusErrorToRangeError(calculationConfig, fe);

    List<Double> rds = testZenithAngles.stream().map(f -> zenithAngleToRangeDistance(calculationConfig.defaultInitialElevation, f) + rde).
      collect(Collectors.toList());

    List<Integer> enc = rds.stream().map(f -> stagePositionToEncoder(controlConfig, rangeDistanceToStagePosition(f))).
      collect(Collectors.toList());

    List<Pair<Pair<Double, Double>, Integer>> tuples = zip(zip(testZenithAngles, rds), enc);
    return tuples.stream().map(TestValue::new).collect(Collectors.toList());
  }

  /**
   * This routine takes the two configs and returns a tuple of ((zenith angle, totalRangeDistance), encoder position)
   * for the range of focus error -20 to 20 in steps of 2 um
   *
   * @param calculationConfig the assembly calculation config provides algorithm parameters
   * @param controlConfig     the control config provides values for mapping stage positions into encoder values
   * @return a Vector of the form (((fe, totalRD), enc), ((fe, totalRD), enc), ...)
   */
  static List<TestValue> calculatedFETestData(TromboneCalculationConfig calculationConfig, TromboneControlConfig controlConfig, Double elevation, Double za) {

    double rangeDistance1 = zenithAngleToRangeDistance(elevation, za);
    List<Double> rangeError = testFocusErrors.stream().map(f -> focusErrorToRangeError(calculationConfig, f)).collect(Collectors.toList());
    List<Double> totalRangeDistance = rangeError.stream().map(f -> f + rangeDistance1).collect(Collectors.toList());

    List<Integer> enc = totalRangeDistance.stream().map(f -> stagePositionToEncoder(controlConfig, rangeDistanceToStagePosition(f))).collect(Collectors.toList());

    List<Pair<Pair<Double, Double>, Integer>> tuples = zip(zip(testFocusErrors, totalRangeDistance), enc);
    return tuples.stream().map(TestValue::new).collect(Collectors.toList());
  }

  static final double minStagePos = rangeDistanceToStagePosition(minTotalRD);
  static final double maxStagePos = rangeDistanceToStagePosition(maxTotalRD);

  // These values take a total el + range to encoder value
  static final List<Pair<Double, Integer>> encoderTestValues = new ArrayList<>();

  static {
    encoderTestValues.add(new Pair<>(80.0, 200));
    encoderTestValues.add(new Pair<>(90.0, 225));
    encoderTestValues.add(new Pair<>(100.0, 305));
    encoderTestValues.add(new Pair<>(110.0, 385));
    encoderTestValues.add(new Pair<>(120.0, 465));
    encoderTestValues.add(new Pair<>(130.0, 545));
    encoderTestValues.add(new Pair<>(140.0, 625));
    encoderTestValues.add(new Pair<>(150.0, 705));
    encoderTestValues.add(new Pair<>(160.0, 785));
    encoderTestValues.add(new Pair<>(170.0, 865));
    encoderTestValues.add(new Pair<>(180.0, 945));
    encoderTestValues.add(new Pair<>(190.0, 1025));
  }
}

File: examples/vslice/src/test/scala/csw/examples/vslice/assembly/AssemblyTestData.scala
package csw.examples.vslice.assembly

import csw.examples.vslice.assembly.AssemblyContext.{TromboneCalculationConfig, TromboneControlConfig}
import csw.services.loc.Connection.AkkaConnection
import csw.services.loc.{ComponentId, ComponentType, Connection}
import csw.services.loc.ConnectionType.AkkaType
import csw.services.pkg.Component.{AssemblyInfo, DoNotRegister, RegisterAndTrackServices}

/**
 * TMT Source Code: 8/12/16.
 */
object AssemblyTestData {

  import Algorithms._

  val testFocusErrors = -20.0 to 20.0 by 4.0
  val testZenithAngles = 0.0 to 60.0 by 5.0

  val hcdId = ComponentId("lgsTromboneHCD", ComponentType.HCD)

  val TestAssemblyInfo = AssemblyInfo(
    "lgsTrombone",
    "nfiraos.ncc.trombone",
    "csw.examples.vslice.assembly.TromboneAssembly",
    RegisterAndTrackServices, Set(AkkaType), Set(AkkaConnection(hcdId))
  )

  val TestCalculationConfig = TromboneCalculationConfig(
    defaultInitialElevation = 95.0,
    focusErrorGain = .75, upperFocusLimit = 20.0, lowerFocusLimit = -20.0,
    zenithFactor = 4
  )

  val TestControlConfig = TromboneControlConfig(
    positionScale = 8.0,
    stageZero = 90.0, minStageEncoder = 225, minEncoderLimit = 200, maxEncoderLimit = 1200
  )

  val TestAssemblyContext = AssemblyContext(TestAssemblyInfo, TestCalculationConfig, TestControlConfig)

  val maxRDError = focusErrorToRangeError(TestCalculationConfig, TestCalculationConfig.upperFocusLimit)
  val maxRD = zenithAngleToRangeDistance(TestCalculationConfig.defaultInitialElevation, testZenithAngles.last)
  val maxTotalRD = maxRDError + zenithAngleToRangeDistance(TestCalculationConfig.defaultInitialElevation, testZenithAngles.last)

  val minRDError = focusErrorToRangeError(TestCalculationConfig, TestCalculationConfig.lowerFocusLimit)
  val minRD = zenithAngleToRangeDistance(TestCalculationConfig.defaultInitialElevation, testZenithAngles.head)
  val minTotalRD = minRDError + zenithAngleToRangeDistance(TestCalculationConfig.defaultInitialElevation, testZenithAngles.head)

  val minReasonableAltitude = 85.0
  val maxReasonableAltitude = 110.0
  val minReasonableRange = minReasonableAltitude
  val maxReasonableRange = 200.0
  val minReasonableStage = minReasonableAltitude
  val maxReasonableStage = 200.0
  val maxReasonableRangeError = 5.0
  val minReasonableRangeError = -5.0

  // These are values for testing the range error based on focus error (fe, fe distance error)
  val feRangeErrorTestValues = testFocusErrors.map(f => (f, focusErrorToRangeError(TestCalculationConfig, f)))

  // This provides "base" range distance based on za only (za, rd)
  val zaRangeDistance = testZenithAngles.map(f => (f, zenithAngleToRangeDistance(TestCalculationConfig.defaultInitialElevation, f)))

  // This produces an Vector of data of form (newRangeDistance, newElevation), (newRangeDistance, newElevation),...) as a function of za at a given focus error
  def newRangeAndElData(fe: Double) = testZenithAngles.map(focusZenithAngleToElevationAndRangeDistance(TestCalculationConfig, TestCalculationConfig.defaultInitialElevation, fe, _))

  // This produces an Vector of data of form (newRangeDistance, newElevation), (newRangeDistance, newElevation),...) as a function of za at a given focus error
  def newZARangeAndElData(za: Double) = testFocusErrors.map(focusZenithAngleToElevationAndRangeDistance(TestCalculationConfig, TestCalculationConfig.defaultInitialElevation, _, za))

  type TestValue = ((Double, Double), Int)

  def getza(t: TestValue): Double = t._1._1

  def gettrd(t: TestValue): Double = t._1._2

  def getenc(t: TestValue): Int = t._2

  /**
   * This routine takes the two configs and returns a tuple of ((zenith angle, totalRangeDistance), encoder position)
   * for the range of zenith angle 0 to 60 in steps of 5 degrees
   *
   * @param calculationConfig the assembly calculation config provides algorithm parameters
   * @param controlConfig the control config provides values for mapping stage positions into encoder values
   * @return a Vector of the form (((za, totalRD), enc), ((za, totalRD), enc), ...)
   */
  def calculatedTestData(calculationConfig: TromboneCalculationConfig, controlConfig: TromboneControlConfig, fe: Double) = {

    val rde = focusErrorToRangeError(calculationConfig, fe)

    val rds = testZenithAngles.map(zenithAngleToRangeDistance(calculationConfig.defaultInitialElevation, _) + rde)

    val enc = rds.map(f => stagePositionToEncoder(controlConfig, rangeDistanceToStagePosition(f)))

    val tuples = testZenithAngles.zip(rds).zip(enc)
    tuples
  }

  /**
   * This routine takes the two configs and returns a tuple of ((zenith angle, totalRangeDistance), encoder position)
   * for the range of focus error -20 to 20 in steps of 2 um
   *
   * @param calculationConfig the assembly calculation config provides algorithm parameters
   * @param controlConfig the control config provides values for mapping stage positions into encoder values
   * @return a Vector of the form (((fe, totalRD), enc), ((fe, totalRD), enc), ...)
   */
  def calculatedFETestData(calculationConfig: TromboneCalculationConfig, controlConfig: TromboneControlConfig, elevation: Double, za: Double) = {

    val rangeDistance1 = zenithAngleToRangeDistance(elevation, za)
    val rangeError = testFocusErrors.map(focusErrorToRangeError(calculationConfig, _))
    val totalRangeDistance = rangeError.map(f => f + rangeDistance1)

    val enc = totalRangeDistance.map(f => stagePositionToEncoder(controlConfig, rangeDistanceToStagePosition(f)))

    val tuples = testFocusErrors.zip(totalRangeDistance).zip(enc)
    tuples
  }

  val minStagePos = rangeDistanceToStagePosition(minTotalRD)
  val maxStagePos = rangeDistanceToStagePosition(maxTotalRD)

  // These values take a total el + range to encoder value
  val encoderTestValues: Vector[(Double, Int)] = Vector(
    (80.0, 200),
    (90.0, 225),
    (100.0, 305),
    (110.0, 385),
    (120.0, 465),
    (130.0, 545),
    (140.0, 625),
    (150.0, 705),
    (160.0, 785),
    (170.0, 865),
    (180.0, 945),
    (190.0, 1025)
  )

}
